File: backend/app/schemas/simulation.py
from pydantic import BaseModel, Field
from typing import List, Optional, Literal
from datetime import datetime

class GeoPosition(BaseModel):
    lat: float = Field(ge=-90.0, le=90.0)
    lng: float = Field(ge=-180.0, le=180.0)

class PilgrimGroupResponse(BaseModel):
    group_id: str
    count: int
    speed_mps: float
    progress: float
    status: Literal["moving", "stopped", "congested"]
    current_edge_id: str
    destination_zone_id: str
    updated_at: datetime
    position: Optional[GeoPosition] = None

class SimulationStateResponse(BaseModel):
    simulation_time: str
    tick: int
    groups: List[PilgrimGroupResponse]

File: frontend/src/types/simulation.ts
export interface GeoPosition {
  lat: number;
  lng: number;
}

export interface PilgrimGroupState {
  group_id: string;
  count: number;
  speed_mps: number;
  progress: number;
  status: 'moving' | 'stopped' | 'congested';
  current_edge_id: string;
  destination_zone_id: string;
  updated_at: string;
  position: GeoPosition | null;
}

export interface CrowdEdgeLoad {
  edge_id: string;
  pilgrim_count: number;
  group_count: number;
  moving_group_count: number;
  stopped_group_count: number;
  congested_group_count: number;
}

export interface CrowdSummary {
  total_pilgrims: number;
  total_groups: number;
  active_edge_count: number;
  moving_group_count: number;
  stopped_group_count: number;
  congested_group_count: number;
  unresolved_group_count: number;
}

export interface CrowdStateSnapshot {
  summary: CrowdSummary;
  edges: CrowdEdgeLoad[];
}

export interface SimulationState {
  simulation_time: string;
  tick: number;
  groups: PilgrimGroupState[];
  crowd: CrowdStateSnapshot;
}
